Better checking for overload function/method declarations.

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59956 [jit] Better checking for overload function/method declarations.**

Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

zhxchen17 committed:
[jit] Better checking for overload function declarations.

Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

zhxchen17 committed:
Update on "[jit] Better checking for overload function declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

zhxchen17 committed:
Update on "[jit] Better checking for overload function declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

gmagogsfm commented on test/test_jit.py:
```diff
@@ -14168,6 +14168,26 @@ def test():
         with self.assertRaisesRegex(Exception, "Parameters not specified"):
             torch.jit.script(test)
 
+        def wrong_decl_body(x: str) -> str:
```
Nit:

Could you create a separate test case for it rather than extending existing test case? Long test case could make test failures less readable

SplitInfinity replied:
Another option is to use [subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests).

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -704,7 +703,17 @@ def copy_torchscript_modifier(orig, new) -> None:
+def _check_overload_body(func):
+    parsed_def = parse_def(func)
+    body = parsed_def.ast.body[0].body
+    if len(body) != 1 or not isinstance(body[0], ast.Pass):
+        msg = "Only `pass` statement can be the body of overload declaration:\n"
+        msg += '\n'.join(parsed_def.source.split("\n")[:3])
+        msg += " <- Expecting `pass` here! "
+        raise RuntimeError(msg)
```
When user hits this error, chances are they do not know how to use `_overload` correctly. Maybe it is worthwhile to go extra mile by showing the correct way of using `_overload` here.

gmagogsfm commented on torch/jit/_script.py:
```diff
@@ -1221,6 +1221,14 @@ def _get_overloads(obj):
 
+    if obj in uncompiled_overloads:
+        sourcelines, file_lineno, filename = _jit_internal.get_source_lines_and_file(obj)
+        raise RuntimeError(
```
Same here, we might want to show complete usage guidance on `_overload` in this case.

gmagogsfm commented on test/test_jit.py:
```diff
@@ -14168,6 +14168,26 @@ def test():
+        def null_overload_driver():
+            return null_overload(0)
+
+        with self.assertRaisesRegex(RuntimeError, 'Implementation for the function ".+" is missing.'):
```
What about the case, say there is a later definition of `null_overload` but its signature mismatches that of its overload declarations?

zhxchen17 replied:
we can add another check to test whether decl and def have the same # of arguments, I guess that's what you mean?

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -1221,6 +1221,14 @@ def _get_overloads(obj):
     if uncompiled_overloads is None:
         return existing_compiled_fns
 
+    if obj in uncompiled_overloads:
```
I am not super familiar with this code. Why does this check imply that there is no definition for the function?

zhxchen17 replied:
If user put "@overload def foo():" as the last toplevel item for overload `foo` i.e. no definition, then jit compiler will recognize that as the "definition" since it's the last defined item in python with that name, so that decl function will appear as `obj` here.

SplitInfinity commented on torch/jit/_script.py:
```diff
@@ -1221,6 +1221,14 @@ def _get_overloads(obj):
+        raise RuntimeError(
+            f'Implementation for the function "{qual_name}" is missing. Please make sure a ' +
+            "definition is provided and defined after all overload declarations.\n" +
+            f'File "{filename}", line {file_lineno}:' + "\n" + ''.join(sourcelines)
```
Is it possible to use `SourceRange` here to highlight source in the same way that we do for exceptions through during compilation of the JIT AST?

zhxchen17 replied:
These checks are performed before we "compile" python asts, so the source ranges are unavailable at this point. Potentially we can move this check into the actual compilation steps, but I find that doesn't make the program simpler, so I'm inclined to a bit straightforward error checking logic rather than a little better error message, assuming this error is likely to be rare.

SplitInfinity reviewed (commented):
Nice!

zhxchen17 commented on torch/_sources.py:
```diff
@@ -0,0 +1,100 @@
+    return SourceContext('', None, 0, 0).make_raw_range(0, 1)
+
+
+ParsedDef = namedtuple('ParsedDef', ['ast', 'ctx', 'source'])
```
didn't know this is a thing! yeah NamedTuple is a lot better.

zhxchen17 committed:
Update on "[jit] Better checking for overload function/method declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

zhxchen17 committed:
Update on "[jit] Better checking for overload function/method declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

zhxchen17 committed:
Update on "[jit] Better checking for overload function/method declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

zhxchen17 committed:
Update on "[jit] Better checking for overload function/method declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

zhxchen17 committed:
Update on "[jit] Better checking for overload function/method declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -716,7 +715,50 @@ def copy_torchscript_modifier(orig, new) -> None:
+def get_overload_no_implementation_error_message(kind, obj):
+    sourcelines, file_lineno, filename = get_source_lines_and_file(obj)
+    return (
+        f'Implementation for the {kind} "{_qualified_name(obj)}" is missing. Please make'
```
```suggestion
        f'Implementation for the {kind} "{_qualified_name(obj)}" is missing. Please make '
```

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -716,7 +715,50 @@ def copy_torchscript_modifier(orig, new) -> None:
+    return (
+        f'Implementation for the {kind} "{_qualified_name(obj)}" is missing. Please make'
+        f'sure a definition is provided and defined after all overload declarations.\n'
+        f'File "{filename}", line {file_lineno}:\n' + ''.join(sourcelines) + _OVERLOAD_EXAMPLE
```
Nit:
I'd add a few empty lines after printing `sourcelines` to prevent `_OVERLOAD_EXAMPLE` from being mixed in user source code.

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -716,7 +715,50 @@ def copy_torchscript_modifier(orig, new) -> None:
+
+    if len(body) != 1 or not (is_pass(body[0]) or is_ellipsis(body[0])):
+        msg = "Only `pass` statement or `...` can be the body of overload declaration:\n"
+        msg += '\n'.join(parsed_def.source.split("\n")[:3])
```
Looks like this is pointing to a segment of problematic code, we have some utility for it:

https://github.com/pytorch/pytorch/blob/35307b131df9d24bfa96103d6061cf14c797ee32/torch/csrc/jit/frontend/source_range.h#L122

Is it possible to use the common `SourceRange` here as well?

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -994,22 +1037,6 @@ def _qualified_name(obj) -> str:
     return module_name + "." + name
 
 
-# Thin wrapper around SourceRangeFactory to store extra metadata
```
Is this code block just moved or moved+modified? 

I think this refactor (along with others you made in this PR) makes sense, just that in general, it is slightly better to put refactoring in a separate PR whenever possible so that each PR has a clear purpose.

zhxchen17 replied:
it's just moved.

gmagogsfm reviewed (approved):
Looks good, just have a few nits.

zhxchen17 committed:
Update on "[jit] Better checking for overload function/method declarations."


Summary:
Issue #50175. Basically two things need to be checked and are lacking currently:
1. Overload declarations should always have a single `pass` statement as the body.
2. There should be always an implementation provided for decls which doesn't
   have the @torch.jit._overload decorator. So in this case we need to check
   whether we are actually compiling a function body with decorator ahead.

Test Plan:
python test/test_jit.py TestScript.test_function_overloads

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29106555](https://our.internmc.facebook.com/intern/diff/D29106555)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29106555).
